src/core/Akka.Tests/Actor/Dispatch/Bug2640Spec.cs

```diff
@@ -0,0 +1,123 @@
+﻿// //-----------------------------------------------------------------------
+// // <copyright file="Bug2640Spec.cs" company="Akka.NET Project">
+// //     Copyright (C) 2009-2019 Lightbend Inc. <http://www.lightbend.com>
+// //     Copyright (C) 2013-2019 .NET Foundation <https://github.com/akkadotnet/akka.net>
+// // </copyright>
+// //-----------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Akka.Actor;
+using Akka.Configuration;
+using Akka.Dispatch;
+using Akka.Routing;
+using Akka.TestKit;
+using FluentAssertions;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Akka.Tests.Actor.Dispatch
+{
+    /// <summary>
+    ///     Bugfix and verification spec for https://github.com/akkadotnet/akka.net/issues/2640
+    ///     Used to verify that the <see cref="ForkJoinExecutor" /> gets properly disposed upon shutdown.
+    /// </summary>
+    public class Bug2640Spec : AkkaSpec
+    {
+        public Bug2640Spec(ITestOutputHelper helper) : base(DispatcherConfig, helper)
+        {
+        }
+
+        private class GetThread
+        {
+            public static readonly GetThread Instance = new GetThread();
+
+            private GetThread()
+            {
+            }
+        }
+
+        /// <summary>
+        ///     Used to pass back data on the current thread.
+        /// </summary>
+        public class ThreadReporterActor : ReceiveActor
+        {
+            public ThreadReporterActor()
+            {
+                Receive<GetThread>(_ => Sender.Tell(Thread.CurrentThread));
+            }
+        }
+
+        public static readonly Config DispatcherConfig = @"
+            myapp{
+                my-pinned-dispatcher {
+                    type = PinnedDispatcher
+                }
+                my-fork-join-dispatcher{
+                    type = ForkJoinDispatcher
+                    throughput = 60
+                    dedicated-thread-pool.thread-count = 4
+                }
+            }
+        ";
+
+        [Fact(DisplayName = "ForkJoinExecutor should terminate all threads upon ActorSystem.Terminate")]
+        public async Task ForkJoinExecutorShouldShutdownUponActorSystemTermination()
+        {
+            var actor = Sys.ActorOf(Props.Create(() => new ThreadReporterActor())
+                .WithDispatcher("myapp.my-fork-join-dispatcher").WithRouter(new RoundRobinPool(4)));
+
+            Dictionary<int, Thread> threads = null;
+            Watch(actor);
+            for (var i = 0; i < 100; i++)
+                actor.Tell(GetThread.Instance);
+
+            threads = ReceiveN(100).Cast<Thread>().GroupBy(x => x.ManagedThreadId)
+                .ToDictionary(x => x.Key, grouping => grouping.First());
+            threads.Count.Should().Be(4, "Expected 4 distinct threads in this example");
+
+            await Sys.Terminate();
+            AwaitAssert(() =>
```

Comment:
Just to make sure I'm completely clear on this, before the fix was implemented this test would have failed due to threads not correctly shutting down?

Reply:
I wasn't able to get this to fail on the build server because it's a racy specification - the memory leak only occurs if the host system is able to terminate the `ActorSystem` faster than 1 second (the default timeout parameter for shutting down dispatchers via the `IScheduler`,) which happened reliably (100% of the time) on my local system during testing. 

However, given that our build servers are running on relatively low-powered Windows Azure VMs, it's possible that the `IScheduler` can win the race. But all we needed to prove that this was the issue was for this test to fail once, which it did.